### ISSUE TITLE
Change in JSON payload in latest version of Grafana

### ISSUE THREAD

**Issue by KiqoCode** [alerta/alerta#1676]
In new version of Grafana they have changed JSON payload format, therefor not compatible with grafana webhook in Alerta.
https://grafana.com/docs/grafana/latest/alerting/unified-alerting/contact-points/

**Comment by jjdmol** [comment]
You can use Grafana's 'Prometheus' webhook instead though, works like a charm!

**Comment by sushovan23** [comment]
So in this case we would need to run alertmanager ? or just use alertmanager and just webhook of alerta for prometheus instead of alertmanager url?

**Comment by jjdmol** [comment]
The latter. You don't need to run an actual Prometheus Alertmanager for this. "AlertManager" in this case describes the _protocol_ for communicating the alarm messages, not the software.

Grafana is capable of emitting its alarms in the AlertManager format, and Alerta is capable of receiving alarms in the AlertManager format. It's just that the webhook in Alerta is called "prometheus", and we have Grafana mimicking Prometheus here, which makes things a bit confusing.

**Comment by mabed-fr** [comment]
Hi,
I confirm that with my grafana in version 10.4.1 the api target /api/webhooks/grafana?api-key=xxx&environment=Production&service=Web&timeout=3600

does not work but with the alertmanager / prometheus target it works almost 100%

Is there a fix planned?

**Comment by mabed-fr** [comment]
v9.0.3 not resolve problem

Payload(json) receive in logs

```
{
  "receiver": "test",
  "status": "firing",
  "alerts": [
    {
      "status": "firing",
      "labels": {
        "alertname": "TestAlert",
        "instance": "Grafana"
      },
      "annotations": {
        "summary": "Notification test"
      },
      "startsAt": "2024-04-12T19:53:04.088799693+02:00",
      "endsAt": "0001-01-01T00:00:00Z",
      "generatorURL": "",
      "fingerprint": "57c6d9296de2ad39",
      "silenceURL": "https://dash.exemple.com/alerting/silence/new?alertmanager=grafana&matcher=alertname%3DTestAlert&matcher=instance%3DGrafana",
      "dashboardURL": "",
      "panelURL": "",
      "values": null,
      "valueString": "[ metric='foo' labels={instance=bar} value=10 ]"
    }
  ],
  "groupLabels": {
    "alertname": "TestAlert",
    "instance": "Grafana"
  },
  "commonLabels": {
    "alertname": "TestAlert",
    "instance": "Grafana"
  },
  "commonAnnotations": {
    "summary": "Notification test"
  },
  "externalURL": "https://dash.exemple.com/",
  "version": "1",
  "groupKey": "test-57c6d9296de2ad39-1712944384",
  "truncatedAlerts": 0,
  "orgId": 1,
  "title": "[FIRING:1] TestAlert Grafana ",
  "state": "alerting",
  "message": "**Firing**\n\nValue: [no value]\nLabels:\n - alertname = TestAlert\n - instance = Grafana\nAnnotations:\n - summary = Notification test\nSilence: https://dash.exemple.com/alerting/silence/new?alertmanager=grafana&matcher=alertname%3DTestAlert&matcher=instance%3DGrafana\n"
}
```

**Comment by hyberdk** [comment]
yes, there have been multiple changes to the grafana alerting (unfortunately), it seems like it has "somewhat" stabilized.. pull requests are welcome to fix the problem ;-)

**Comment by ShadowJonathan** [comment]
I'm probably encountering this problem; I'm seeing that I'm getting HTTP 400 errors when attempting to use the grafana webhook

**Comment by ShadowJonathan** [comment]
AlertManager v1 api has been deprecated; https://prometheus.io/docs/alerting/latest/clients/

**Comment by Hipska** [comment]
Does someone know which version of Grafana webhooks is supported by Alerta?

**Comment by Hipska** [comment]
Okay; I found a solution/workaround for Grafana not having the legacy alerting anymore in latest versions.


Since Grafana v12.0 there is an option to configure a [custom payload](https://grafana.com/docs/grafana/v12.0/alerting/configure-notifications/manage-contact-points/integrations/webhook-notifier/#custom-payload) for webhook notifications.

The following could be set as payload template:

```go
{{ define "webhook.legacy.payload" }}
	{{- with (index .Alerts 0) }}
		{{- coll.Dict 
			"title" .Annotations.summary
			"ruleId" .Fingerprint
			"ruleName" (or .Labels.rulename .Labels.alertname)
			"ruleUrl" .GeneratorURL
			"state" (tmpl.Inline `{{ if eq .Status "resolved" }}ok{{ else }}alerting{{ end }}` .)
			"message" (temple.Inline `{{ if .Labels.rulename }}{{ .Labels.alertname }}{{ else }}{{ .Annotations.description }}{{ end }}` .)
			"evalMatches" (coll.Slice (coll.Dict 
				"metric" .Labels.name
				"value" .Values
				"tags" (.Labels.Remove (stringSlice "grafana_folder" "alertname" "name"))
			))
			"tags" (.Annotations.Remove (stringSlice "description"))
		| data.ToJSONPretty " "}}
	{{- end }}
{{- end }}
```

Be sure to **disable grouping** in the alert notification or notification policy options, only the first alert can be processed.

In my example, the alert rule instances have a `name` label, which is used as the `metric` name in `evalMatches` (which is then used for `resource` in Alerta). Feel free to adjust to match your setup. 


PS, I also created a template which is compatible with `/api/alert` endpoint, let me know if interested.

**Comment by satterly** [comment]
Closing -- Grafana webhook payloads have changed significantly since this was filed. The current webhook handler supports recent Grafana versions.

**Comment by Hipska** [comment]
@satterly which commits are the fix for this? I think that the current Grafana hook still only supports legacy alerts, not current Grafana alerts?
